### PR TITLE
Update tests and setup as 3.11 is officially released

### PR DESCRIPTION
**Pull request by oscargus**
#### Brief description of what is fixed or changed

I guess we can claim that we support 3.11 now?

#### Release Notes

NO ENTRY

### CODE REVIEW

File: .github/workflows/runtests.yml [hunk 1]
```diff
@@ -172,11 +172,12 @@ jobs:
...
-        python-version: ['3.9', '3.10', 'pypy-3.8']
+        python-version: ['3.9', '3.10', '3.11', 'pypy-3.8']
         experimental: [false]
-        include:
-          - python-version: '3.11.0-alpha - 3.11'
-            experimental: true
+        # Maybe use this to add 3.12 when the time comes:
+        #include:
+        #  - python-version: '3.12.0-alpha - 3.12'
+        #    experimental: true
```
Comment (oscarbenjamin) [inline .github/workflows/runtests.yml:180]: Can this not be done now?

Comment (oscargus) [reply .github/workflows/runtests.yml:180]: It doesn't look like it is possible to install?

Comment (oscargus) [reply .github/workflows/runtests.yml:180]: https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md

Comment (oscargus) [reply .github/workflows/runtests.yml:180]: PyPy
2.7.18 [PyPy 7.3.9]
3.6.12 [PyPy 7.3.3]
3.7.13 [PyPy 7.3.9]
3.8.13 [PyPy 7.3.9]
3.9.12 [PyPy 7.3.9]
Python
2.7.18
3.6.15
3.7.15
3.8.14
3.9.15
3.10.8
3.11.0

Comment (oscargus) [reply .github/workflows/runtests.yml:180]: Not sure where the required tests is set, but it should be changed....

```
doctests (3.11.0-alpha - 3.11, false) Expected — Waiting for status to be reported
Required
tests1 (3.11.0-alpha - 3.11, false) Expected — Waiting for status to be reported
Required
tests2 (3.11.0-alpha - 3.11, false) Expected — Waiting for status to be reported
```

Comment (oscarbenjamin) [reply .github/workflows/runtests.yml:180]: Maybe this is holding it up: https://github.com/actions/python-versions/pull/197

Comment (oscarbenjamin) [reply .github/workflows/runtests.yml:180]: > Not sure where the required tests is set, but it should be changed....

It's in the repo settings in the branch protection rules. I've just removed those required jobs. New jobs will have to be set as required when the list is finalised.

**Comment by oscarbenjamin** [comment]
> I guess we can claim that we support 3.11 now?

Yes. We've been testing 3.11 in CI for a long time. Some upstream bugs were found because of testing the 3.11 prereleases so we should do the same for 3.12.

**Comment by oscargus** [comment]
> > I guess we can claim that we support 3.11 now?
> 
> Yes. We've been testing 3.11 in CI for a long time. Some upstream bugs were found because of testing the 3.11 prereleases so we should do the same for 3.12.

Agreed. I'm also thinking of creating a weekly job that installs "all" nightly versions of the optional dependencies (numpy, scipy, and matplotlib if nothing else) and runs the optional tests with those to get a heads up in case something breaks.

**Comment by oscarbenjamin** [comment]
> Agreed. I'm also thinking of creating a weekly job that installs "all" nightly versions of the optional dependencies (numpy, scipy, and matplotlib if nothing else) and runs the optional tests with those to get a heads up in case something breaks.

Currently that's what the 3.11 prerelease job does but it conflates two different things because it uses latest prerelease CPython along with bleeding edge git numpy, scipy etc. The problem with that is that some of the upstream dependencies like scipy etc don't start testing with prerelease CPython until late in its dev cycle so many of those dependencies might be uninstallable and I had to keep fiddling with which things could or could not be installed.

I think it would be better to have two separate things:

- An optional dependency job that uses a slightly older CPython e.g. 3.10 or 3.9 but installs bleeding edge versions of the most used optional dependencies (mpmath, gmpy2, numpy, scipy, etc).
- Another set of jobs that just runs the ordinary tests and doctests against the latest prerelease of CPython.

Finally another nice to have would be a bleeding edge sphinx job. It's quite common that releases of sphinx break something in the sympy docs so it would be good to pick those things up before the sphinx release in case they can be fixed there.

**Comment by oscargus** [comment]
This removes the bleeding edge matplotlib and numpy builds (as there are 3.11 wheels). Hence, the benefit of running those separately (and indeed not necessarily on bleeding edge Python).

Sphinx can also make sense for sure.

**Pull request merged**